lib/translator/write.cc

```diff
@@ -0,0 +1,249 @@
+    return status_code;
+  }
+
+  status_code = LegacyWrite(nvme_wrapper, allocation, nsid,
+                            transfer_length_pages, buffer_out);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  uint8_t pr_info;
+  status_code = BuildPRInfo(wrprotect, pr_info);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+  nvme_wrapper.cmd.cdw[2] = htoll(BuildCdw12(transfer_length, pr_info, fua));
+
+  return status_code;
+}
+
+}  // namespace
+
+StatusCode Write6ToNvme(Span<const uint8_t> scsi_cmd,
+                        NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                        uint32_t nsid, uint32_t page_size, uint32_t lba_size,
+                        Span<const uint8_t> buffer_out) {
+  scsi::Write6Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write6 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  // A TRANSFER LENGTH field set to zero specifies that 256 logical blocks shall
+  // be written. Any other value specifies the number of logical blocks that
+  // shall be written (Section 3.59) of
+  // https://www.seagate.com/files/staticfiles/support/docs/manual/Interface%20manuals/100293068j.pdf
+  uint16_t updated_transfer_length =
+      (write_cmd.transfer_length == 0) ? 256 : write_cmd.transfer_length;
+
+  uint32_t transfer_length_pages = 0;
+  StatusCode status_code = GetTransferLengthPages(
+      transfer_length_pages, updated_transfer_length, page_size, lba_size);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  status_code = LegacyWrite(nvme_wrapper, allocation, nsid,
+                            transfer_length_pages, buffer_out);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  // cdw10 Starting lba bits 31:00
+  uint32_t host_endian_lba = (write_cmd.logical_block_address_1 << 16) +
+                             ntohs(write_cmd.logical_block_address_2);
+
+  nvme_wrapper.cmd.cdw[0] = htoll(host_endian_lba);
+  nvme_wrapper.cmd.cdw[2] =
+      htoll(static_cast<uint32_t>(updated_transfer_length - 1));
+  return status_code;
+}
+
+StatusCode Write10ToNvme(Span<const uint8_t> scsi_cmd,
+                         NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                         uint32_t nsid, uint32_t page_size, uint32_t lba_size,
+                         Span<const uint8_t> buffer_out) {
+  scsi::Write10Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write10 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  StatusCode status_code = Write(
+      write_cmd.fua, write_cmd.wr_protect, ntohs(write_cmd.transfer_length),
+      nvme_wrapper, allocation, nsid, page_size, lba_size, buffer_out);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_wrapper.cmd.cdw[0] = __bswap_32(write_cmd.logical_block_address);
```

Comment:
Should we be using htonl or similar here instead, for better portability?

Reply:
I believe the value here is copied straight from the SCSI buffer so it is in big endian. So bswap is the more appropriate choice

Reply:
Makes sense. Is there a non-underscore-prefixed variant we can use? Usually double underscores indicate implementation details that aren't meant to be used.

Reply:
There is and we should make the change to use that. Somehow this was missed in previous commits to master so I'll open an issue to make the change https://github.com/googleinterns/scsi2nvme/issues/191